tests/functional/u/used/used_before_assignment_nonlocal.py

```diff
@@ -133,3 +135,31 @@ def inner():
                 def callback():
                     pass
     outer()
+
+
+def nonlocal_in_distant_outer_frame_fail(callback, condition_a, condition_b):
+    """Nonlocal declared in outer frame, usage and definition immediately enclosed
+    in intermediate frame.
+    """
+    def outer():
+        nonlocal callback
+        def intermediate():
+            if condition_a:
+                def inner():
+                    callback()  # [possibly-used-before-assignment]
+                inner()
+            else:
+                if condition_b:
+                    def callback():
+                        pass
+        intermediate()
+    outer()
+
+
+def nonlocal_after_bad_usage_fail():
+    """Nonlocal declared after used-before-assignment."""
+    num = 1
+    def outer():
+        num = num + 1  # [used-before-assignment]
+        nonlocal num
+    outer()

```

Comment:
this is a new (and deliberate) message introduced from #10034, we probably want to raise an error when the nonlocal statement is evaluated after the use and assignment